### PR TITLE
🐛[bento-base-carousel] remove window fallback in debouncedResetScrollR…

### LINKED ISSUE #36990: [bento-base-carousel] Does not currently render server side with NextJS
### Description

While implementing the carousel in a NextJS app I got the following error.

```
ReferenceError: window is not defined
```

It appears that we're using the window to set an interval to automatically increment the slides on the carousel. However when rendering on the server side there is no window context, so it throws an error.

### Reproduction Steps

Start a nextjs app

```
npx create-next-app@latest
```

change directories to the new project and install the carousel

```
npm install @bentoproject/base-carousel
```

implement a basic carousel in the `pages/index.js`

Run nextjs app

```
npm run dev
```

Get error.

### Relevant Logs

_No response_

### Browser(s) Affected

_No response_

### OS(s) Affected

_No response_

### Device(s) Affected

_No response_

### AMP Version Affected

1.2110212130.2

### PR DESCRIPTION
**Pull request by dethstrobe**
…eferencePoint in scroller.js

resolves #36990

### CODE REVIEW

File: extensions/amp-base-carousel/1.0/scroller.js [hunk 1]
```diff
@@ -192,7 +192,7 @@ function ScrollerWithRef(
   // Trigger render by setting the resting index to the current scroll state.
   const debouncedResetScrollReferencePoint = useMemo(() => {
     // Use local window if possible.
-    const win = containerRef.current ? getWin(containerRef.current) : window;
+    const win = getWin(containerRef.current);
     return debounce(
       win,
```
Comment (jridgewell) [inline extensions/amp-base-carousel/1.0/scroller.js:197]: While debounce itself is pure, the function that it returns is bound to the window instance when it was created. When that function is invoked, it'll try to dereference a `win.Date`, and if `win` is null, it'll throw.

What I think needs to happen is `debouncedResetScrollReferencePoint` should itself be put inside a `useLayoutEffect`. We can use a `useRef` to ensure stability of the function instance.

**Review by jridgewell — CHANGES_REQUESTED** [review]

File: extensions/amp-base-carousel/1.0/scroller.js (hunk 1 above)
Comment (dethstrobe) [reply extensions/amp-base-carousel/1.0/scroller.js:197]: I do believe that `useMemo` is the correct hook to use here. I feel like creating a ref and using a hook to update the ref is redundant.

What I think we can do is create a condition if win is undefined we can just execute an empty function that doesn't do anything.

Comment (jridgewell) [reply extensions/amp-base-carousel/1.0/scroller.js:197]: As long as `win` is closed over, and `useMemo` doesn't take into account that `containerRef.current` has changed from `null` to any node, then we'll have a broken debounce that can't actually execute. You can hoist this out and add `window` as part of the memo:

```js
const win = getWin(containerRef.current);
const debouncedResetScrollReferencePoint = useMemo(() => {
  return debounce(win, …);
}, [restingIndex, setRestingIndex, win]);
```

(Also, debounce fires when the timer ends, but we may have cleaned up the memo here. That's a bug crying for a uniform `useDebounceCallback` helper)

File: extensions/amp-base-carousel/1.0/scroller.js [hunk 2]
```diff
@@ -191,23 +191,24 @@ function ScrollerWithRef(
...
+              currentIndex.current === restingIndex
+            ) {
+              return;
+            }
+            setRestingIndex(currentIndex.current);
+          },
+          RESET_SCROLL_REFERENCE_POINT_WAIT_MS
+        )
+      : () => undefined;
   }, [restingIndex, setRestingIndex]);
```
Comment (jridgewell) [inline extensions/amp-base-carousel/1.0/scroller.js:212]: You need to account for `containerRef.current`/`win` to bust the memo when the node is mounted.

**Review by jridgewell — APPROVED** [review]

**Pull request closed without merging**